src/common/Result.h

```diff
@@ -0,0 +1,240 @@
+// Copyright 2018 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef COMMON_RESULT_H_
+#define COMMON_RESULT_H_
+
+#include "common/Assert.h"
+#include "common/Compiler.h"
+
+#include <cstdint>
+
+// Result<T, E> is the following sum type (Haskell notation):
+//
+//      data Result T E = Success T | Error E | Empty
+//
+// It is meant to be used as the return type of functions that might fail. The reason for the Empty
+// case is that a Result should never be discarded, only destructured (its error or success moved
+// out) or moved into a different Result. The Empty case tags Results that have been moved out and
+// Result's destructor should ASSERT on it being Empty.
+//
+// Since C++ doesn't have efficient sum types for the special cases we care about, we provide
+// template specializations for them.
+
+template <typename T, typename E>
+class Result;
+
+// The interface of Result<T, E> shoud look like the following.
+//  public:
+//    Result(T success);
+//    Result(E error);
+//
+//    Result(Result<T, E>&& other);
+//    Result<T, E>&& operator=(Result<T, E>&& other);
+//
+//    ~Result();
+//
+//    bool IsError() const;
+//    bool IsSuccess() const;
+//
+//    T&& AcquireSuccess();
+//    E&& AcquireError();
+
+// Specialization of Result for returning errors only via pointers. It is basically a pointer
+// where nullptr is both Success and Empty.
+template <typename E>
+class NXT_NO_DISCARD Result<void, E*> {
+  public:
+    Result();
+    Result(E* error);
+
+    Result(Result<void, E*>&& other);
+    Result<void, E*>& operator=(Result<void, E>&& other);
+
+    ~Result();
+
+    bool IsError() const;
+    bool IsSuccess() const;
+
+    void AcquireSuccess();
+    E* AcquireError();
+
+  private:
+    E* mError = nullptr;
+};
+
+// Specialization of Result when both the error an success are pointers. It is implemented as a
+// tagged pointer. The tag for Success is 0 so that returning the value is fastest.
```

Comment:
This would mean that `IsSuccess` would need to check for the tag and for nullptr. An alignment virtually every struct has an alignment of 4 so I don't think we need to reduce the number of bits.

Reply:
Ok, I agree.